src/request/component-models/LoggingComponentModel.ts

```diff
@@ -1,39 +1,171 @@
-    };
+export class LoggingComponentModel extends ComponentModel implements ILoggingComponentModel {
+    private _levels: LoggingLevelModel[];
+    private _messages: ILogMessageModel[];
 
-    private _messages: ILogMessage[];
+    public constructor(private _glimpse: IGlimpse, private _requestDetailStore: IRequestDetailStore, private _messageProcessor) {
+        super();
+    }
+
+    public get levels(): ILoggingLevelModel[] {
+        return this._levels;
+    }
 
-    public get messages() {
-        return this._messages;
+    public get totalMessageCount(): number {
+        return this._messages.length;
     }
 
     public init(request) {
-        const allMessages = messageProcessor.getTypeStucture(request, LoggingComponentModel.options);
+        const options = {
+            'log-write': this._messageProcessor.getTypeMessageList
+        };
+
+        const allMessages = this._messageProcessor.getTypeStucture(request, options);
 
         if (allMessages) {
             this._messages = _(allMessages.logWrite)
                 .sortBy<IMessageEnvelope<ILogMessage>>('ordinal')
-                .map(message => message.payload)
+                .map((message, index) => new LogMessageModel(message, index + 1))
+                .value();
+
+            const levels: { [key: string]: ILogMessageModel[] } = {};
+
+            _.defaults(levels, _.groupBy(this._messages, message => message.level), {
+                Debug: [],
+                Verbose: [],
+                Information: [],
+                Warning: [],
+                Error: [],
+                Critical: []
+            });
+
+            this._levels = _(levels)
+                .transform(
+                    (result: LoggingLevelModel[], messages, level) => {
+                        result.push(new LoggingLevelModel(level, messages.length));
+                    },
+                    [])
+                .sortBy(level => LoggingComponentModel.getOrderOfLevel(level.level))
                 .value();
         }
         else {
+            this._levels = [];
             this._messages = [];
         };
     }
+
+    public getMessages(): ILogMessageModel[] {
+        const state = this._requestDetailStore.getState().logging.filter;
+
+        let filteredMessages = _(this._messages);
+
+        if (state) {
+            _.filter(this._levels, level => {
+                return state[level.level] === false;
+            })
+            .forEach(level => {
+                filteredMessages = filteredMessages.filter(message => message.level !== level.level);
+            });
+        }
+
+        return filteredMessages.value();
+    }
+
+    public isShown(level: ILoggingLevelModel): boolean {
+        const state = this._requestDetailStore.getState().logging.filter;
+
+        return state === undefined || state[level.level] !== false;
+    }
+
+    public toggleAll(): void {
+        const state = _.clone(this._requestDetailStore.getState().logging.filter);
```

Comment:
It's cloned because, in Redux (which our architecture parallels in many ways, and to which we might decide to migrate), state is always replaced rather than mutated.  This new state is published/dispatched and picked up by the store, which will then cause the UI to refresh and reflect the new state.